### DISCUSSION TITLE
/hyndsight/rolling-forecasts/

### DISCUSSION THREAD

**Discussion by robjhyndman** [robjhyndman/robjhyndman.com#193] [Announcements]
Variations on rolling forecasts 
 https://robjhyndman.com/hyndsight/rolling-forecasts/

**Comment by robjhyndman** [comment]
```header
Originally posted by Alberto Santini on 2014-07-16 09:11:29
````
hsales is a ts object.

start and end parameters can be either a single number or a vector of
two integers, which specify a natural time unit and a (1-based) number
of samples into the time unit.

window(hsales,end=c(1989, 12)) \# more readable? :)

**Comment by robjhyndman** [comment]
```header
Originally posted by Louis Pinget on 2014-07-16 14:38:41
````
Thank you very much Professor. I\'ll try to use a Multi-step forecasts
with re-estimation.\
but, I have some questions. My goal is not to validate my model. I want
to forecast some new data with an updating ARIMA. So, if I take the same
data than you, my problem is that: I have data from 1973 to 1989 (train)
but I haven't data from 1990 to 1995 (test). So I need to construct &quot;x&quot;
with data from 1973 to 1989 and estimated data from 1989.99 + (i-1)/12
(with i in 1:67). I don't know how I could do it on R.

**Reply by robjhyndman** [reply]
```header
Originally posted on 2014-07-16 23:11:04
````
Use the forecast command.

**Comment by robjhyndman** [comment]
```header
Originally posted by Spencer on 2014-08-12 14:46:47
````
is it possible to implement the multi-step w/o re-estimation with
bats/tbats? If so are there examples of doing this anywhere in your blog
or in other documentation?\
Thanks!

**Reply by robjhyndman** [reply]
```header
Originally posted on 2014-08-12 23:49:24
````
Not yet. Re-fitting a bats/tbats model to new data is on the wish list
for new features.
[https://github.com/robjhynd\...](https://github.com/robjhyndman/forecast/issues/23)

**Reply by robjhyndman** [reply]
```header
Originally posted by Spencer on 2014-08-13 14:41:57
````
great, looking forward to it. Thanks for the response and for the very
useful R package and other documentation you have provided!

**Comment by robjhyndman** [comment]
```header
Originally posted by Lisa on 2014-10-10 12:29:47
````
I would like to do multi-step forecasts with re-estimation for bivariate
VAR models. Could you tell me how I need to modify the code then?\
Thank you!

**Comment by robjhyndman** [comment]
```header
Originally posted by Affy on 2015-01-08 01:39:34
````
Dear Professor,

Does &quot;forecast&quot; command work for multiple linear regression as well?

**Reply by robjhyndman** [reply]
```header
Originally posted on 2015-01-08 04:27:36
````
Yes

**Comment by robjhyndman** [comment]
```header
Originally posted by Affy on 2015-01-08 07:16:48
````
Thanks a lot professor for your reply.

**Comment by robjhyndman** [comment]
```header
Originally posted by Ummah Salma on 2015-08-16 15:05:41
````
I am trying to do one-step forecast for workers\' remittances data for
Bangladesh. The data span is from 1980 4th quarter to 2014 4th quarter.
here are my R codes:

Arima:

train \&lt;- window(lremit,end=2007,4)

fit10 \&lt;- auto.arima(train)

refit \&lt;- Arima(lremit,model=fit10)

fc\&lt;- window(fitted(refit),start=2008,1)

fc

ets:

fit11 \&lt;- ets(train)

refit2 \&lt;- ets(lremit,model=fit11)

fc2\&lt;- window(fitted(refit2),start=2008,1)

fc2

my problems are:

1\. I want to measure accuracy for these one-step forecasts, but can not
do it properly.

2\. In case of ets() it shows the error for fc2,&quot;Error: object \'fc2\'
not found&quot;

3\. There are other error messages both for arima and ets like ,

\&gt; train \&lt;- window(lremit,end=2007,4)

Warning message: In window.default(x, \...) : \'start\' value not
changed

\&gt; fc\&lt;- window(fitted(refit),start=2008,1)

Error in window.default(x, \...) : \'start\' cannot be after \'end\'

Although these messages are not hampering my work, but I want to know\
the better way to handle this.

Thank you.

**Reply by robjhyndman** [reply]
```header
Originally posted on 2015-08-16 21:22:35
````
You need to specify end=c(2007,4) and start=c(2008,1)

**Reply by robjhyndman** [reply]
```header
Originally posted by Ummah Salma on 2015-08-17 10:32:05
````
Professor,

Thank you for the reply, it solved the error messages problem. After
one-step forecast, I need the code to measure accuracy.

**Comment by robjhyndman** [comment]
```header
Originally posted by Yang-hui Chang on 2016-01-21 01:46:51
````
Would anyone could tell me :

Why air.model can be used in out-of-sample multi-step forecasts or
onestep forecasts??\
If one\'s data is yearly data ,may the onestep forecasts is one year
head forecast?

\# Fit model to first few years of AirPassengers data

air.model \&lt;- Arima(window(AirPassengers,end=1956+11/12),order=c(0,1,1),

seasonal=list(order=c(0,1,1),period=12),lambda=0)

plot(forecast(air.model,h=48))

lines(AirPassengers)

\# Apply fitted model to later data

air.model2 \&lt;- Arima(window(AirPassengers,start=1957),model=air.model)

\# Forecast accuracy measures on the log scale.

\# in-sample one-step forecasts.

accuracy(air.model)

\# out-of-sample one-step forecasts.

accuracy(air.model2)

\# out-of-sample multi-step forecasts

accuracy(forecast(air.model,h=48,lambda=NULL),

log(window(AirPassengers,start=1957)))

**Comment by robjhyndman** [comment]
```header
Originally posted by Yang-hui Chang on 2016-01-21 01:53:37
````
Can anyone tell me how to install \'fpp\'??

**Comment by robjhyndman** [comment]
```header
Originally posted by Yang-hui Chang on 2016-01-21 01:56:40
````
there is no package called 'fpp'

**Reply by robjhyndman** [reply]
```header
Originally posted by Peng on 2016-01-26 02:52:15
````
install.packages(&quot;fpp&quot;)

library(fpp)

**Comment by robjhyndman** [comment]
```header
Originally posted by Malcolm Vu on 2016-04-01 05:34:39
````
I was wondering, why the forecasting done for the testing set starts on
the 5th month of the testing set?\
Can I change it to :

n \&lt;- length(test)

and

fc \&lt;- ts(numeric(n), start=1990, freq=12)

Thank you in advance.

**Comment by robjhyndman** [comment]
```header
Originally posted by Ilias on 2017-03-12 20:34:22
````
Dear Rob, first of all thanks for the spreading of your amazing
knowledge. This site and the otext book are a real treasure.\
I\'d like to know how the above code should be modified for a
multivariate ARIMA. Do you then also need to apply a window on the
predictor matrix?\
Thanks in advance

**Reply by robjhyndman** [reply]
```header
Originally posted on 2017-03-12 21:05:35
````
For a multivariate ARIMA, the code is virtually identical and there is
no predictor matrix. For a regression with ARIMA errors, you will need
to use a similar window on the predictor matrix (argument xreg)

**Reply by robjhyndman** [reply]
```header
Originally posted by Ilias on 2017-03-13 14:00:21
````
First of all, thanks for the quick reply! So if I apply ARIMAX to the
Multi-step forecasts with re-estimation, I\'d get something like this:

\# Multi-step, re-estimation\
h \&lt;- 5\
train \&lt;- window(hsales,end=1989.99)\
test \&lt;- window(hsales,start=1990)\
pred_train \&lt;- window(predictor_matrix, end = 1989.99)

n \&lt;- length(test) - h + 1\
fit \&lt;- auto.arima(train, xreg = pred_train)\
order \&lt;- arimaorder(fit)\
fcmat \&lt;- matrix(0, nrow=n, ncol=h)\
for(i in 1:n)\
{\
x \&lt;- window(hsales, end=1989.99 + (i-1)/12)\
pred_train \&lt;- window(predictor_matrix, end = 1989.99 + (i-1)/12)\
refit \&lt;- Arima(x, xreg= pred_train)\
fcmat\[i,\] \&lt;- forecast(refit, h=h, xreg= pred_train)\$mean\
}

Kind regards

**Comment by robjhyndman** [comment]
```header
Originally posted by ispanyolca on 2017-03-17 18:56:40
````
Dear Profesor, I have sent mail to you about this issue.\
In the annex, you can see rolling and recursive methods, basically.\
Your scripts are different from these?\
Sincerely\
[https://uploads.disquscdn.c\...](https://uploads.disquscdn.com/images/0bf7518c07758e03a2f87a8484301b514c07a9099bd7bb9c12b0bd7f57ad784c.png)

**Comment by robjhyndman** [comment]
```header
Originally posted by Robert Boer on 2017-06-22 10:07:03
````
Dear prof. Hyndman, thank you very much for the explanation on this!

In your introduction, you specify that the code can easily be adapted to
other univariate time series models. However, I\'m struggling a bit with
doing so for the tslm function.

My model takes quite long to run so that\'s why I chose to evaluate my
ARIMA and nnetar models without re-estimating of the models. I\'ve been
able to evaluate both with one-step forecasts and multi-step forecasts.
Now I\'m trying to compare this with the performance of the Time-Series
Linear Model, using the tslm function. I\'m struggling a bit with the
step in which the refit object is created, as simply using the code
\'refitTSLM \&lt;- tslm(x_msts \~ season, model = fitTSLM)\' returns the
following error:

Error in if (model) z\$model \&lt;- mf :\
argument is not interpretable as logical\
In addition: Warning message:\
In if (model) z\$model \&lt;- mf :\
the condition has length \&gt; 1 and only the first element will be used

It seems that the \'model = \' argument does not exist within the tslm
function. Is there another argument which I can use to specify that for
the refitted model the tslm function should make use of the model
specified in the fitTSLM object?

Many thanks in advance!

**Reply by robjhyndman** [reply]
```header
Originally posted on 2017-06-22 11:12:37
````
The tslm function does not have a model argument. If you want to apply a
fitted model to some new data, just use forecast() with the relevant
inputs. It is just a regression model, so you don\'t need to use
recursion.

**Reply by robjhyndman** [reply]
```header
Originally posted by Robert Boer on 2017-06-22 12:48:46
````
Thanks for your answer. So if I understand it correctly, as fitted
values are simply one-step forecasts, I can fit a tslm on the full data
and then test the data for one-step forecasts using the following code:

fitTSLM \&lt;- tslm(Full_ts \~ season)\
forecastTSLM \&lt;- window(fitted(fitTSLM), start = c(12, 364)) \# c(12,
364) is where test set starts)

And for 6-step ahead forecasts I use:

\# \-\-\-\-- 6-step forecast without re-estimation\
h \&lt;- 6\
train \&lt;- window(Full_ts, end = c(12, 363))\
test \&lt;- window(Full_ts, start = c(12, 364))\
n \&lt;- length(test) - h + 1\
fcTSLM \&lt;- ts(numeric(n), start = c(12, 369), freq = 1008)

for(i in 1:n)\
{\
x \&lt;- window(Full_ts, end = c(12, 363) + (i-1)/1008)\
fitTSLM \&lt;- tslm(x \~ season)\
fcTSLM\[i\] \&lt;- forecast(fitTSLM, h = h)\$mean\[h\]\
}

Does that make sense?

**Reply by robjhyndman** [reply]
```header
Originally posted on 2017-06-22 22:43:43
````
The problem with this code is that you are using the test data to fit
the model. Perhaps try using the tsCV function which wraps up a lot of
the work for you.

**Comment by robjhyndman** [comment]
```header
Originally posted by Vitor Curtis on 2017-10-16 19:26:47
````
Does anyone know what is the best approach to forecast with rolling
mean, but returning a forecast object? I wrote the following function
that can return a timeseries object using rollmean function, but I
cannot convert it to forecast object. (the code is probably ugly because
I am learning R)

forecast_mm\_ \&lt;- function(y, h=52, meanfunction=rollmean, window=5) {\
y \&lt;- as.ts(y)\
n \&lt;- length(y)\
f \&lt;- y\*NA\
freq \&lt;- frequency(y)

\# first prediction considers h and window\
first \&lt;- max(window, h)+1\
fc \&lt;- try(suppressWarnings(meanfunction(y, k=window, na.pad=TRUE,
align=\'right\')), silent=TRUE)\
f\[first:n\] \&lt;- fc\[window:(length(fc)-h-1)\]

\# expand to new periods\
for (i in 1:h) {\
f \&lt;- c(f, mean( f\[(length(f)-window+1):length(f)\] ))\
}\
f \&lt;- (f\[(length(f)-h+1):length(f)\])

\# msts object\
newStart \&lt;- end(y)\
newStart\[2\] \&lt;- newStart\[2\]+1\
if (newStart\[2\] \&gt; freq) {\
newStart\[2\] \&lt;- 1\
newStart\[1\] \&lt;- newStart\[1\]+1\
}\
return(ts(f, start=newStart, frequency=freq))\
}

forecast_mm\_(ts, h=10)

**Comment by robjhyndman** [comment]
```header
Originally posted by Telis on 2018-05-29 10:42:48
````
Hi Professor and congrats for your great work.\
Suppose that we want to run a forecasting race between the best Arima
and the best ETS model (based on their out-of-sample RMSFE), where the
winning model-algorithm (with the lowest RMSFE) at the end of the day
lets say that it will forecast the sales of a product. In this case,
would you prefer a rolling scheme which re-estimates and re-selects the
best Arima and ETS models OR selecting the models just once at the
beginning (based on their in-sample fit - AIC) and then re-estimating
them by extending the training data in each time period - iteration?
Which one would you choose?\
Much appreciated.

**Reply by robjhyndman** [reply]
```header
Originally posted on 2018-05-29 22:20:32
````
If you re-select a model, then you are testing which model selection
algorithm is best. If you estimate once only, you are testing which
pre-specified model is best. The first is more realistic in comparing
what happens in practice. But you might want to do the second in some
circumstances.

**Reply by robjhyndman** [reply]
```header
Originally posted by Samy on 2018-09-26 06:34:11
````
Hi Prof.. I too had same doubt regarding choosing model by re-select &amp;
re-estimate or select once and re-estimate again because as per our
practice, we first select and model and test to determine performance
and then use same model with full data to forecast future values.
Re-selecting &amp; re-estimating again after performance check will fail our
objective to test model on out of sample set. Plz help.. thanks

**Comment by robjhyndman** [comment]
```header
Originally posted by amrapalli karan on 2018-08-31 04:09:13
````
Hi Professor and thank you for your work.\
I have auction data that is aggregated every Saturday and is taken for a
period of 2014,07 to 2018,03. The data set has moving averages of
currency exchange rates, Baltic dry index and Crude Oil Price along with
the weekly aggregated Bid Price. I am suppose to forecast the Bid Price
for the next 52 weeks. Can rolling forecasts be applicable for
multivariate forecasts as in my case? If yes then once I have divided
the data into test and train, what should the \'\'xreg=&quot; argument be
and how to proceed? Thanks in advance.

**Comment by varmamanoj** [comment]
Hello sir, I want to weather any researcher related to this method is published by you? Also want to know if you have done work related to testing set size determination. Please reply

**Reply by robjhyndman** [reply]
This blog post is published by me. The test set should be the same size as the forecast horizon of interest, if possible.

**Comment by nurainmg** [comment]
Dear Prof

I have developed an LSTM model using keras tensorflow package. I used  multiple data sequences, each consisting 12 input values with 1 target point for training and validation. Upon getting the desired hyperparameters, I would now want to make a recursive forecast for 12 data points in the future. How do i go about this? Is there any function in R that i can use to make the 12 points recursive forecast?or do i need to use a loop method?

Thank you in advance. I appreciate your input

**Reply by robjhyndman** [reply]
It needs a loop.
